File: exercises/chapter7.cpp
#include <RayTracerChallenge/datastructures/matrix.hpp>
#include <RayTracerChallenge/datastructures/vec4.hpp>
#include <RayTracerChallenge/objects/camera.hpp>
#include <RayTracerChallenge/objects/canvas.hpp>
#include <RayTracerChallenge/objects/colour.hpp>
#include <RayTracerChallenge/objects/sphere.hpp>
#include <RayTracerChallenge/objects/world.hpp>
#include <fstream>
#include <numbers>

int main() {
  rtc::World world{};

  auto floor = rtc::sphere();
  auto floorMaterial = floor.getMaterial();
  floorMaterial.colour = rtc::Colour{1, 0.9, 0.9};
  floorMaterial.specular = 0;
  floor.setTransformationMatrix(rtc::scalingMatrix(10, 0.01f, 10));
  world.objects.push_back(std::make_unique<rtc::Sphere>(floor));

  auto leftWall = rtc::sphere();
  leftWall.setTransformationMatrix(
      rtc::translationMatrix(0, 0, 5) *
      rtc::rotationMatrixY(-std::numbers::pi_v<float> / 4) *
      rtc::rotationMatrixX(std::numbers::pi_v<float> / 2) *
      rtc::scalingMatrix(10, 0.01f, 10));
  leftWall.setMaterial(floor.getMaterial());
  world.objects.push_back(std::make_unique<rtc::Sphere>(leftWall));

  auto rightWall = rtc::sphere();
  rightWall.setTransformationMatrix(
      rtc::translationMatrix(0, 0, 5) *
      rtc::rotationMatrixY(std::numbers::pi_v<float> / 4) *
      rtc::rotationMatrixX(std::numbers::pi_v<float> / 2) *
      rtc::scalingMatrix(10, 0.01f, 10));
  rightWall.setMaterial(floor.getMaterial());
  world.objects.push_back(std::make_unique<rtc::Sphere>(rightWall));

  auto middle = rtc::sphere();
  middle.setTransformationMatrix(rtc::translationMatrix(-0.5f, 1, 0.5f));
  auto middleMaterial = middle.getMaterial();
  middleMaterial.colour = rtc::Colour{0.1f, 1, 0.5f};
  middleMaterial.diffuse = 0.7f;
  middleMaterial.specular = 0.3f;
  middle.setMaterial(middleMaterial);
  world.objects.push_back(std::make_unique<rtc::Sphere>(middle));

  auto right = rtc::sphere();
  right.setTransformationMatrix(rtc::translationMatrix(1.5f, 0.5f, -0.5f) *
                                rtc::scalingMatrix(0.5f, 0.5f, 0.5f));
  auto rightMaterial = right.getMaterial();
  rightMaterial.colour = rtc::Colour{0.5f, 1, 0.1f};
  rightMaterial.diffuse = 0.7f;
  rightMaterial.specular = 0.3f;
  right.setMaterial(rightMaterial);
  world.objects.push_back(std::make_unique<rtc::Sphere>(right));

  auto left = rtc::sphere();
  left.setTransformationMatrix(rtc::translationMatrix(-1.5f, 0.33f, -0.75f) *
                               rtc::scalingMatrix(0.33f, 0.33f, 0.33f));
  auto leftMaterial = left.getMaterial();
  leftMaterial.colour = rtc::Colour{1, 0.8f, 0.1f};
  leftMaterial.diffuse = 0.7f;
  leftMaterial.specular = 0.3f;
  left.setMaterial(leftMaterial);
  world.objects.push_back(std::make_unique<rtc::Sphere>(left));

  world.light = rtc::Light{rtc::Colour{1, 1, 1}, rtc::point(-10, 10, -10)};

  rtc::Camera camera{200, 200, std::numbers::pi_v<float> / 3};
  camera.setTransform(rtc::viewTransform(
      rtc::point(0, 1.5f, -5), rtc::point(0, 1, 0), rtc::vector(0, 1, 0)));

  const rtc::Canvas canvas = camera.render(world);
  std::ofstream file;
  file.open("Chapter7.ppm");
  file << canvas.toPPM();
  file.close();
  return 0;
}

File: RTC.Console/Exercises/Chapter7.cs
using RTC.Datastructures;
using RTC.Objects;
using RTC.Primitives;

namespace RTC.Console.Exercises;

public class Chapter7 : IChapterExercise
{
    public void Run()
    {
        var world = new World();

        var floor = new Sphere();
        floor.Material = floor.Material with
        {
            Colour = new Colour(1, 0.9d, 0.9d),
            Specular = 0
        };
        floor.Transformation = Matrix.ScalingMatrix(10, 0.01d, 10);
        world.Shapes.Add(floor);
        
        var leftWall = new Sphere
        {
            Transformation = Matrix.TranslationMatrix(0, 0, 5) * 
                             Matrix.RotationYMatrix(-Math.PI / 4) *
                             Matrix.RotationXMatrix(Math.PI / 2) *
                             Matrix.ScalingMatrix(10, 0.01d, 10)
        };
        world.Shapes.Add(leftWall);
        
        var rightWall = new Sphere
        {
            Transformation = Matrix.TranslationMatrix(0, 0, 5) * 
                             Matrix.RotationYMatrix(Math.PI / 4) *
                             Matrix.RotationXMatrix(Math.PI / 2) *
                             Matrix.ScalingMatrix(10, 0.01d, 10)
        };
        world.Shapes.Add(rightWall);

        var middle = new Sphere
        {
            Transformation = Matrix.TranslationMatrix(-0.5d, 1, 0.5d)
        };
        middle.Material = middle.Material with
        {
            Colour = new Colour(0.1d, 1, 0.5d),
            Diffuse = 0.7d,
            Specular = 0.3d
        };
        world.Shapes.Add(middle);

        var right = new Sphere
        {
            Transformation = Matrix.TranslationMatrix(1.5d, 0.5d, -0.5d) * Matrix.ScalingMatrix(0.5d, 0.5d, 0.5d)
        };
        right.Material = right.Material with
        {
            Colour = new Colour(0.5d, 1, 0.1d),
            Diffuse = 0.7d,
            Specular = 0.3d
        };
        world.Shapes.Add(right);

        var left = new Sphere
        {
            Transformation = Matrix.TranslationMatrix(-1.5d, 0.33d, -0.75d) * Matrix.ScalingMatrix(0.33d, 0.33d, 0.33d)
        };
        left.Material = left.Material with
        {
            Colour = new Colour(1, 0.8d, 0.1d),
            Diffuse = 0.7d,
            Specular = 0.3d
        };
        world.Shapes.Add(left);
        
        world.Lights.Add(new PointLight(Colour.White, Vec4.Point(-10, 10, -10)));

        var camera = new Camera(300, 300, Math.PI / 3)
        {
            Transformation = Matrix.ViewTransform(Vec4.Point(0, 1.5, -5), Vec4.Point(0, 1, 0), Vec4.Vector(0, 1, 0))
        };
        
        var canvas = camera.Render(world);
        canvas.SavePng("Chapter7.png");
    }
}